                     Evalucación Final Módulo 1 "Python"
                        Mónica García Jiménez

In [484]:
# Creamos tres estructuras de datos
inventory = []
customers = {}
total_sales = 0.0

In [485]:

inventory.append({'name': 'shirt', 'price': 20, 'quantity': 40})
inventory.append({'name': 'pants', 'price': 30, 'quantity': 30})

In [486]:

customers['customer1'] = {'email': 'customer1@email.com', 'shopping': []}
customers['customer2'] = {'email': 'customer2@email.com', 'shopping': []}


In [487]:

def agregaadd_products(name, price, quantity = inventory): # Con esta función podemos añadir un producto a nuestro inventario, comprobando si existe.
 
    for product in inventory:# Compruebe si el producto ya existe en la lista.
        if product ["name"]== name: 
            print(f"El producto {name} ya está registrado. Actualizando los datos...")
            product['quantity'] += quantity # Actualizar cantidad si está registrado
            return 
      
    new_product = {"name" : name, "price": price, "quantity": quantity}
    # Si el producto no está registrado, se crea un diccionario actualizado
    inventory.append(new_product)
    print(f"Producto {name} agregado con éxito.")


In [488]:
# Ejemplo de uso
agregaadd_products("socks", 5, 10)
agregaadd_products("skirts", 40, 25)
agregaadd_products("shirt", 20,  40)
agregaadd_products("pants",  30,  30)

Producto socks agregado con éxito.
Producto skirts agregado con éxito.
El producto shirt ya está registrado. Actualizando los datos...
El producto pants ya está registrado. Actualizando los datos...


In [490]:

def view_inventory():# Con la siguiente función mostraremos el inventario
    for product in inventory: # Muestra el inventario
            if len(inventory) == 0:
                print("El inventario se encuentra vacio")
            else: # Si no esta vacio, nos muestra los productos mas a fondo
                print(f"product: {product['name']}, price: {product['price']}, quantity: {product['quantity']}")
                




In [492]:
def search_product(name): # Podemos observar el inventario
    for product in inventory:
        if product['name'].lower() == name.lower():  # Compara sin importar mayúsculas o minúsculas
            print(f"name: {product['name']}, price: {product['price']}, quantity: {product['quantity']}")
            return product# Sale de la función después de encontrar el producto
              
    print(f"El producto '{name}' no está en el inventario.")  # Mensaje si no se encuentra el producto
             


In [493]:
#Ejemplo de uso
search_product("shirt")
search_product("pants")
search_product("glasses")


name: shirt, price: 20, quantity: 80
name: pants, price: 30, quantity: 60
El producto 'glasses' no está en el inventario.


In [494]:

def update_stock(name, quantity):# Con esta función actualizams el stock 
    
    for product in inventory: #recorre el inventario buscando el producto
        if product ["name"] == name.lower():
            
            
            print(f"Stock actualizado: {product['name']} ahora tiene {product['quantity']} unidades.")
            return
# Verificamos que no se intente establecer una cantidad negativo
        
    print(f"El producto '{name}' no está en el inventario.")
            


In [495]:
# Ejemplo de uso
update_stock("shirt", 50)
update_stock("socks",20)
update_stock ("glasses", 10)


Stock actualizado: shirt ahora tiene 80 unidades.
Stock actualizado: socks ahora tiene 10 unidades.
El producto 'glasses' no está en el inventario.


In [501]:
def remove_product(name): # Con esta función eliminamos un producto del inventario
    
    for product in inventory:
        if product ["name"] == name.lower(): 
            inventory.pop(product) # Elimina el elemento que le indiquemos
            print(f"Stock actualizado: {product['name']} ahora tiene {product['quantity']} unidades.")
            return

    # Verificamos que no se intente establecer una cantidad negativo   
    print(f"El producto '{name}' no se encuentra en el inventario.")

In [505]:
# Llama a la función para comprobar si está en el inventario
remove_product("pants")


El producto 'pants' no se encuentra en el inventario.


In [508]:
def calculate_value_inventory(): # Con esta función nos permite calcular el valor total del inventario
    total_value = 0.0
    for product in inventory:
        price = product ['price']
        quantity = product ['quantity']
        total_value += price * quantity # Se calcula el total
    return total_value
    value_inventory = calculate_value_inventory(inventory)
print(f'El valor total del inventario es: {value_inventory} euros')
           

El valor total del inventario es: 850 euros


In [509]:
def make_purchase(): 

    customer_purchases = [] # Creamos una lista de la compra vacía para el cliente.

    cart= {}
    while True: # Usamos el bucle While para seguir comprando
        view_inventory () #mostrar inventario    
        product =input("Ingrese el nombre del producto que desea comprar (o 'salir' para finalizar): ").lower()# Pregunta al cliente que producto quiere (por nombre).

        if product == 'salir':
            break
        
        if product not in inventory:
            print("Producto no disponible. Intente nuevamente.")
            continue
        
        quantity = int(input(f"Ingrese la cantidad de {product} que desea comprar:")) # Pregunta la cantidad.
         
        if quantity > inventory [product]["quantity"]: # Comporbamos si tenemos suficiente stock
            print(f"No hay suficiente cantidad de {product}. Disponibles: {inventory[product]['quantity']}.")
            continue
         # Actualizar carrito e inventario
        if product in cart:
            cart[product] += quantity
        else:
            cart[product] = quantity
        
        inventory[product]["quantity"] -= quantity
        print(f"Agregado {quantity} de {product} al carrito.")

        # Calcular el costo total
        total_sales = 0.0
        for product in customer_purchases:
            total_sales = total_sales + product['price'] * product['quantity']

        print(f'Valor total: {total_sales}')
        return customer_purchases

In [510]:
# Llamar a la función
make_purchase()

product: shirt, price: 20, quantity: 80
product: socks, price: 5, quantity: 10
product: skirts, price: 40, quantity: 25
Producto no disponible. Intente nuevamente.
product: shirt, price: 20, quantity: 80
product: socks, price: 5, quantity: 10
product: skirts, price: 40, quantity: 25
Producto no disponible. Intente nuevamente.
product: shirt, price: 20, quantity: 80
product: socks, price: 5, quantity: 10
product: skirts, price: 40, quantity: 25


In [511]:

def process_payment(): # Con esta función realizamos el proceso de pago
    try:
        # Solicita la cantidad total de la compra
        total_purchase = float(input("Ingresa el total de la compra: "))
        
        # Solicita la cantidad de pago
        payment_amount = float(input("Ingresa la cantidad de pago: "))
        
        # Calcula el cambio
        if payment_amount < total_purchase:
            print("Error: La cantidad de pago es insuficiente.")
        else:
            change = payment_amount - total_purchase
            print(f"Pago exitoso. Su cambio es: {change:.2f}")
    
    except ValueError:
        print("Error: Por favor, ingrese un número válido.")

In [512]:
# Llamar a la función para procesar el pago
process_payment()

Pago exitoso. Su cambio es: 30.00


In [564]:
customers = []

In [566]:

customers = {}

def add_customer(name, email):# Esta función nos sirve para agregar un cliente a nuestro diccionario de clientes
    
    if name in customers:# Verifica si el cliente ya existe
        print("El cliente ya está registrado.")
    else:
        # Agrega el cliente al diccionario
        customers[name] = email
        print(f"Cliente agregado: {name} con correo {email}")
        print(f"Cliente agregado con éxito")

In [569]:
customers

{'Monica Garcia': 'monicagar@gmail.com',
 'Juan Bonilla': 'juanbm@gmail.com',
 'Lorena Garcia': 'lorenagm@gmail.com'}

In [568]:
# Ejemplo de uso
add_customer("Monica Garcia" , "monicagar@gmail.com"),
add_customer("Juan Bonilla" , "juanbm@gmail.com"),
add_customer("Lorena Garcia" , "lorenagm@gmail.com") 

Cliente agregado: Monica Garcia con correo monicagar@gmail.com
Cliente agregado con éxito
Cliente agregado: Juan Bonilla con correo juanbm@gmail.com
Cliente agregado con éxito
Cliente agregado: Lorena Garcia con correo lorenagm@gmail.com
Cliente agregado con éxito


In [570]:


def view_customers(customers): # Con esta función vemos la lista de clietes registrados
    # Comprobamos si la lista de clientes tiene algún cliente añadido.
    if len(customers) == 0:
        print('\nLa lista de clientes está vacía.')  # La lista está vacía.
    else:
        # Si no, recorremos la base de datos de clientes.
        print(f"\nClientes:")
        print(f"-------------\n")
        for name, email in customers.items():
            print(f"name: {name}, email: {email}")  # Imprimimos los detalles del cliente.


In [571]:
# Mostrar todos los clientes
view_customers(customers)


Clientes:
-------------

name: Monica Garcia, email: monicagar@gmail.com
name: Juan Bonilla, email: juanbm@gmail.com
name: Lorena Garcia, email: lorenagm@gmail.com


In [572]:
customers

{'Monica Garcia': 'monicagar@gmail.com',
 'Juan Bonilla': 'juanbm@gmail.com',
 'Lorena Garcia': 'lorenagm@gmail.com'}

In [573]:


def register_purchase(name_customer, shopping_cart):# Con la siguiente función, podemos registrar una compra
    
    if name_customer not in customers: # Verificamos si el cliente está en el diccionario.
        print(f"No se puede realizar la acción porque el cliente '{name_customer}' no está en el diccionario.")
        return
    # Calcular el total de la compra
    total_purchase = 0
    for product in shopping_cart():
        total_purchase += product["price"] * product ["quantity"]
        purchase = {"products": shopping_cart,"total": total_purchase}
    # Actualizar las ventas totales del cliente y de la tienda
    customers[name_customer]["shopping_cart"].append(purchase)
    customers [name_customer]["total_sales"] += total_purchase
    print(f"Compra registrada para {name_customer} Total de la compra: €{total_purchase}.")

In [574]:
# Ejemplo de uso
cart1 = [{'name': 'shirt', 'price': 20, 'cantidad': 7}, {'name': 'shirt', 'price': 20, 'cantidad': 4}]
cart2 = [{'name': 'socks', 'price': 5, 'cantidad': 2}, {'name': 'skirts', 'price': 40, 'cantidad': 8}]

In [576]:
register_purchase('Lorena', cart2)

In [450]:
def view_customer_purchases(name_customer): # permite ver el historial de compras de un cliente:
    
    if name_customer in customers:# Verificamos si el cliente existe en el diccionario
        shopping = customers[name_customer]["shopping"]
        if shopping:
            print(f"Historial de compras de {name_customer}:")
            for buy in shopping:
                print(f"- product: {buy['product']}, price: {buy['price']}, "
                      f"quantity: {buy['quantity']}, total: {buy['total']}")
        else:
            print(f"{name_customer} no ha realizado ninguna compra.")
    else:
        print(f"El cliente {name_customer} no se encuentra en el sistema.")



In [583]:
# Esta función calcula el total de las ventas:
def calculate_total_sales():
    total_sales = 0.0

    # Obtenemos el historial de compras.
    for name in customers:
        total_sales += customers[name]['total_purchase']
    
    return total_sales

## 4.Execution

In [ ]:
# Add products to inventory
agregaadd_products('shirt', 20, 40)
agregaadd_products('pants', 30, 30)
agregaadd_products('pants', 20, 20) 
agregaadd_products("socks", 5, 10)

In [ ]:
# View inventory
print("Inventario:")
view_inventory()

# Search for a product in inventory
search_product("shirt")

# Update the stock of a product
update_stock("shirt", 50)

# Remove a product from inventory
remove_product("pants")

# Calculate total inventory value
calculate_value_inventory()

# Make a purchase
make_purchase()

# Procesar un pago
process_payment()

# Agregar clientes
add_customer("Monica Garcia" , "monicagar@gmail.com"),
add_customer("Juan Bonilla" , "juanbm@gmail.com"),
add_customer("Lorena Garcia" , "lorenagm@gmail.com")
add_customer("Monica Garcia" ,  "monicagar@gmail.com")

# Ver lista de clientes
view_customers()

# Registrar una compra para un cliente
cart1 = [{'name': 'shirt', 'price': 20, 'cantidad': 7}, {'name': 'shirt', 'price': 20, 'cantidad': 4}]
cart2 = [{'name': 'socks', 'price': 5, 'cantidad': 2}, {'name': 'skirts', 'price': 40, 'cantidad': 8}]

# Calcular ventas totales de la tienda

calculate_total_sales() 


Inventario vacio:

El inventario está vacío
La lista de clientes esta vacia

-----------------------------------------------------------
Añadimos productos al inventario y mostramos como se ve:

name: shirt, price: 20,  quantity: 80
name: pants, price: 30,  quantity: 60
name: socks, price:  5,  quantity: 10
name: skirts, price: 40, quantity: 25

------------------------------------------------------------
Buscamos el producto:

name: shirt, price: 20, quantity: 80
name: pants, price: 30, quantity: 60
El producto 'glasses' no está en el inventario.

------------------------------------------------------------
Actualizamos el stock:

Stock actualizado: shirt ahora tiene 40 unidades.
Stock actualizado: socks ahora tiene 10 unidades.
El producto 'glasses' no está en el inventario.

------------------------------------------------------------
Eliminamos un producto:

El producto 'pants' no se encuentra en el inventario.

------------------------------------------------------------
Calcular el valor del inventario:

El valor total del inventario es: 850 euros

------------------------------------------------------------
Clientes:
Resumen compra:
Pago exitoso. Su cambio es: 30.00
Cliente agregado: Monica Garcia con correo monicagar@gmail.com
Cliente agregado con éxito
Cliente agregado: Juan Bonilla con correo juanbm@gmail.com
Cliente agregado con éxito
Cliente agregado: Lorena Garcia con correo lorenagm@gmail.com
Cliente agregado con éxito
El cliente no se encuentra en el diccionario
